### ISSUE TITLE
question about sampling cuda kernel 

### ISSUE THREAD

**Issue by gyin-ai** [NVIDIA/FasterTransformer#97]
can I ask what those following steps mean in the topk_kernel? Thanks.

1. why do we need ` total.u = __expf(total.u - s_max);`? 
total.u seems to be the logits. 

2. what does s_sum mean here? why do we need the `rand_num = (float)curand_uniform(curandstate + blockIdx.x) * s_sum;`?

```
if(tid == 0) 
        {
            s_id[ite] = total.p;
            s_val[total.p] = -MAX_T_VAL;
            total.u = __expf(total.u - s_max);
	        s_val2[total.p] = (T)total.u;
	        s_sum += total.u;
        }
        __syncthreads();
    }
    if(tid == 0)
    {
        rand_num = (float)curand_uniform(curandstate + blockIdx.x) * s_sum;
        for(int i = 0; i < k; i++)
        {
            rand_num = rand_num - (float)s_val2[s_id[i]];
            if(rand_num <= 0.0f)
            {
                ids[batch_id] = topk_tmp_id_buf[batch_id * size + s_id[i]]  % vocab_size;
                break;
            }
        }
```
https://github.com/NVIDIA/FasterTransformer/blob/93b7751bcbb4b7b29ce647b646baa6414790d4d0/fastertransformer/cuda/topk_kernels.cu#L379

**Comment by byshiue** [comment]
1. It is a common idea to prevent overflow during computing the probability.
2. It is used to sample from the top k tokens.
